paprika/restraints/plumed.py

```diff
@@ -0,0 +1,348 @@
+            self.group_atoms = {}
+
+            # Parse each restraint in the list
+            for restraint in self.restraint_list:
+                # Skip restraint if the target or force constant is not defined.
+                # Example: wall restraints only used during the attach phase.
+                try:
+                    target = restraint.phase[phase]["targets"][window]
+                    force_constant = (
+                        restraint.phase[phase]["force_constants"][window]
+                        * self.k_factor
+                    )
+                except TypeError:
+                    continue
+
+                # Convert list to comma-separated string
+                atom_index = self._get_atom_indices(restraint)
+                atom_string = ",".join(map(str, atom_index))
+
+                # Determine collective variable type
+                colvar_type = "distance"
+                if len(atom_index) == 3:
+                    colvar_type = "angle"
+                    target *= _PI_ / 180.0
+                elif len(atom_index) == 4:
+                    colvar_type = "torsion"
+                    target *= _PI_ / 180.0
+
+                # Determine bias type for this restraint
+                bias_type = get_bias_potential_type(restraint, phase, window)
+                if bias_type is "harmonic":
+                    bias_type = "restraint"
+
+                # Append string to lists
+                cv_list.append(
+                    f"c{cv_index}: {colvar_type.upper()} ATOMS={atom_string} NOPBC\n"
+                )
+                bias_list.append(
+                    f"{bias_type.upper()} ARG=c{cv_index} AT={target:.4f} KAPPA={force_constant:.2f}\n"
+                )
+
+                # Increment colvar index
+                cv_index += 1
+
+            # Write collective variables to file
+            self._write_colvar_to_file(windows, cv_list, bias_list)
+
+    def _write_colvar_to_file(self, window, cv_list, bias_list):
+        with open(os.path.join(self.path, window, self.file_name), "a") as file:
+            if len(self.group_atoms) != 0:
+                file.write("# Center groups\n")
+                for key, value in self.group_atoms.items():
+                    file.write(f"{key}: COM ATOMS={value}\n")
+
+            file.write("# Collective variables\n")
+            for line in cv_list:
+                file.write(line)
+
+            file.write("# Bias potentials\n")
+            for line in bias_list:
+                file.write(line)
+
+    def _get_atom_indices(self, restraint):
+        # Check atom index setting
+        index_shift = 0
+        if not restraint.amber_index:
+            index_shift = 1
+            logger.info("Atom indices starts from 0 --> shifting indices by 1.")
+
+        # Collect DAT atom indices
+        atom_index = []
+        for i in range(4):
+            ii = i + 1
+
+            if not eval(f"restraint.index{ii}"):
```

Comment:
This... seems fishy. I'm not entirely clear what you're testing here, but I think you can just see if `restraint.index3` is `None` (or something like that), right?

Reply:
> if restraint.index3 is None (or something like that), right?
True, this is a better logic statement.

The code for this function is actually very similar to that from the function `amber_restraint_line` https://github.com/slochower/pAPRika/blob/96219fea95ced11f9785857bdfda665ea1acc8ae/paprika/restraints/amber.py#L8
but `codeclimate` was complaining about the code complexity so I had to resort to a for loop and `eval` statement.

Reply:
I hate to suggest more work, but in this case, I'd expect the earlier version you had (if it is closer to `amber_restraint_line`) is the way to go (without `eval`) and `codeclimate` can be damned. Looks like `openff-evaluator` isn't even using `codeclimate` anymore, so perhaps we can get rid of that check too. I think I added it when OpenFF was just starting out and some projects had it. My suggestion is:

1. Revert this back to not use `eval`
2. (Separate PR): remove `codeclimate` and setup GH actions to run `black`, `isort`, and `flake8` like so: https://github.com/openforcefield/openff-evaluator/blob/master/.github/workflows/lint.yaml 

What do you think?

Reply:
Okay, getting rid of `codeclimate` sounds good. Seems to be rejecting a lot of my code. I'll revert this code back to it's original before `eval`.

Yeah, what Evaluator has is pretty good. I'll do a separate PR for this.

Reply:
I've reverted the code to it's original